de/nmichael/efa/data/efacloud/TableBuilder.java

```diff
@@ -0,0 +1,520 @@
+            "Statistics;FILTERBYPERSONTEXT;DATA_STRING;FilterByPersonText",   //
+            "Statistics;SHOWLOGBOOKFIELDS;DATA_LIST_STRING;ShowLogbookFields",   //
+            "Statistics;AGGRDISTANCEBARSIZE;DATA_INTEGER;AggregationDistanceBarSize",   //
+            "Statistics;AGGRROWDISTANCEBARSIZE;DATA_INTEGER;AggregationRowDistanceBarSize",   //
+            "Statistics;AGGRCOXDISTANCEBARSIZE;DATA_INTEGER;AggregationCoxDistanceBarSize",   //
+            "Statistics;AGGRSESSIONSBARSIZE;DATA_INTEGER;AggregationSessionsBarSize",   //
+            "Statistics;AGGRAVGDISTBARSIZE;DATA_INTEGER;AggregationAvgDistanceBarSize",   //
+            "Statistics;AGGRDURATIONBARSIZE;DATA_INTEGER;AggregationDurationBarSize",   //
+            "Statistics;AGGRSPEEDBARSIZE;DATA_INTEGER;AggregationSpeedBarSize",   //
+            "Statistics;OPTIONTRUNCATEDIST;DATA_BOOLEAN;OptionTruncateDistance",   //
+
+            // Statistics data fields which need stricter limitation because of MySQL row length
+            // using 1024 characters of DATA_LIST_UUID instead of 2048 chars of DATA_LIST_STRING
+            "Statistics;FILTERGENDER;DATA_LIST_UUID;FilterGender",   //
+            "Statistics;FILTERSESSIONTYPE;DATA_LIST_UUID;FilterSessionType",   //
+            "Statistics;FILTERBOATTYPE;DATA_LIST_UUID;FilterBoatType",   //
+            "Statistics;FILTERBOATSEATS;DATA_LIST_UUID;FilterBoatSeats",   //
+            "Statistics;FILTERBOATRIGGING;DATA_LIST_UUID;FilterBoatRigging",   //
+            "Statistics;FILTERBOATCOXING;DATA_LIST_UUID;FilterBoatCoxing",   //
+            "Statistics;FILTERBOATOWNER;DATA_LIST_UUID;FilterBoatOwner",   //
+
+            // Data fields common for versionized tables. They are already added
+            // within the metadata definitions an only listed for completeness.
+            "$versionized;VALIDFROM;DATA_LONGINT;ValidFrom",  //
+            "$versionized;INVALIDFROM;DATA_LONGINT;InvalidFrom",  //
+            "$versionized;INVISIBLE;DATA_BOOLEAN;Invisible",  //
+            "$versionized;DELETED;DATA_BOOLEAN;Deleted", //
+
+            // Data fields common for all tables. They are already added
+            // within the metadata definitions an only listed for completeness.
+            "$allTables;CHANGECOUNT;DATA_LONGINT;ChangeCount",  //
+            "$allTables;LASTMODIFIED;DATA_LONGINT;LastModified"};
+
+    // cache to hold all special fields for checking when building the tables
+    private final HashMap<String, RecordFieldDefinition> specialFields = new HashMap<>();
+    // the complete structure of all record types, i. e. MySQL tables.
+    public HashMap<String, RecordTypeDefinition> recordTypes = new HashMap<>();
+    // when starting the application, data will be synchronized by loading them from the server
+    // to show the load progress, the GUI context is needed. It can be put here
+    public EfaBaseFrame guiBaseFrameOnAppLoading = null;
```

Comment:
This is currently only being set from EfaBaseFrame (under some conditions which aren't clear to me), but not from EfaBoathouseFrame. Does it need to be set from both?

Reply:
EfaBoathouseFrame does write any structure information into the efaCloud DB, but only records. It therefore does not need that HashMap. Nonetheless I will have a look. Inconsistencies shall be avoided, I agrre.

Reply:
guiBaseFrameOnAppLoading is used by the de.nmichael.efa.data.efacloud.EfaCloudSynch synchronization process as parentDialog in the synchronization progress display by de.nmichael.efa.gui.ProgressDialog. Now the synchronization can only be triggered from the BaseFrame menu, as it was not implemented by me in the efaBoatHouse menu. I just didn't deal with it for now. If it shall be usable also in that context via admin login, this needs still to be done.